File: 150.evalRPN.go
package main

import "strconv"

func evalRPN(tokens []string) int {
	stack := make([]int, len(tokens))
	top := -1

	for i := 0; i < len(tokens); i++ {
		s := tokens[i]

		if s == "+" {
			stack[top-1] += stack[top]
			top--
		} else if s == "-" {
			stack[top-1] -= stack[top]
			top--
		} else if s == "*" {
			stack[top-1] *= stack[top]
			top--
		} else if s == "/" {
			stack[top-1] /= stack[top]
			top--
		} else {
			top++
			stack[top], _ = strconv.Atoi(s)
		}
	}

	if top != 0 {
		return -1
	}

	return stack[0]
}

File: 150.evalRPN.cpp
#include "shdstd-cpp.h"

using namespace std;

class Solution {
public:
	int evalRPN(vector<string> &tokens) {
		stack<int> st;

		for (int i = 0; i < tokens.size(); ++i) {
			string s = tokens[i];
			if (s != "+" && s != "-" && s != "*" && s != "/") {
				st.push(stoi(s));
				continue;
			}

			int b = st.top();
			st.pop();
			int a = st.top();
			st.pop();

			if (s == "+") {
				st.push(a + b);
			} else if (s == "-") {
				st.push(a - b);
			} else if (s == "*") {
				st.push(a * b);
			} else {
				st.push(a / b);
			}
		}

		return st.top();
	}
};
